docs/proposals/kubelet-systemd.md

```diff
@@ -0,0 +1,266 @@
+
+A `slice` corresponds to an inner-node in the `cgroup` file-system hierarchy.
+
+For example, the `system.slice` is represented as follows:
+
+`/sys/fs/cgroup/<controller>/system.slice`
+
+A `slice` is nested in the hierarchy by its naming convention.
+
+For example, the `system-foo.slice` is represented as follows:
+
+`/sys/fs/cgroup/<controller>/system.slice/system-foo.slice/`
+
+A `service` or `scope` corresponds to leaf nodes in the `cgroup` file-system hierarchy.
+
+For example, if the `docker.service` is associated with the `system.slice`, it is
+represented as follows:
+
+`/sys/fs/cgroup/<controller>/system.slice/docker.service/`
+
+To demonstrate the use of `scope` units using the `docker` container runtime, if a
+user launches a container via `docker run -m 100M busybox`, a `scope` will be created
+because the process was not launched by `systemd` itself.  The `scope` is parented by
+the `slice` associated with the launching daemon.
+
+For example:
+
+`/sys/fs/cgroup/<controller>/system.slice/docker-<container-id>.scope`
+
+`systemd` defines a set of slices.  By default, service and scope units are placed in
+`system.slice`, virtual machines and containers registered with `systemd-machined` are
+found in `machine.slice`, and user sessions handled by `systemd-logind` in `user.slice`.
+
+## kubelet cgroup driver
+
+The current implementation of the `kubelet` reads and writes to the `cgroup` file-system
+directly during initial bootstrapping in ways that are not appropriate for `systemd`
+systems.
+
+This proposal proposes that the `kubelet` accept a new flag:
+
+* `--cgroup-driver="":` - cgroup driver used by the kubelet.
+Possible values: `cgroupfs` or `systemd`.
+
+The `ContainerManager` interface should support an implementation per driver option.
+
+Shared code across implementations should be split into a common location.
+
+By default, the `kubelet` should default `--cgroup-driver` to `systemd` on `systemd` distributions.
+
+## kubelet cgroup bootstrapping under systemd
+
+The following sections outline how a `systemd` system supports local node accounting
+requirements.
+
+### Node capacity
+
+The `kubelet` will continue to interface with `cAdvisor` to determine node capacity.
+
+### System reserved
+
+The node may set aside a set of designated resources for non-Kubernetes components.
+
+The `kubelet` accepts a flag of the following form to reserve compute resources
+for non-Kubernetes components:
+
+* `--system-reserved=:` - A set of `ResourceName`=`ResourceQuantity` pairs that
+describe resources reserved for host daemons.
+
+The `kubelet` does not enforce `--system-reserved`, but the ability to distinguish
+the static reservation from observed usage is important for node accounting.
+
+The `kubelet` will accept a new flag in a `systemd` environment to support accounting
+system-reserved resources.
+
+* `--slice-system="system.slice"` - Comma-delimited set of slices reserved for
```

Comment:
I may change this and state that if the Kubelet is launched to use the systemd cgroup driver, the system-container-name glad we have today will accept input that conforms to slice naming convention.

Originally, I thought that we may have to pass additional slices (possibly machine.slice) but I am not sure that is required at this point.

It would be easier if we could use the same existing flag but just expect systemd conventions on input.